lib/token-store.js

```diff
@@ -8,196 +8,1110 @@
+    }
+
+    /** @inheritdoc */
+    get lastTokens() {
+        return this.tokens.slice(this.firstIndex, this.lastIndex + 1);
+    }
+}
+
+/**
+ * The cursor which iterates tokens and comments.
+ * @private
+ */
+class TokenCommentCursor extends BaseCursor {
+
     /**
-     * Gets a number of tokens that precede a given node or token in the token
-     * stream.
-     * @param {(ASTNode|Token)} node The AST node or token.
-     * @param {int} [beforeCount=0] The number of tokens before the node or
-     *     token to retrieve.
-     * @returns {Token[]} Array of objects representing tokens.
+     * Initializes this cursor.
+     * @param {Token[]} tokens - The array of tokens.
+     * @param {Comment[]} comments - The array of comments.
+     * @param {Object} imap - The map from locations to indices in `this.tokens`.
+     * @param {number} start - The start location of the iteration range.
+     * @param {number} end - The end location of the iteration range.
      */
-    api.getTokensBefore = function(node, beforeCount) {
-        const first = starts[node.range[0]];
+    constructor(tokens, comments, imap, start, end) {
+        super(tokens, comments, imap, start, end);
+        this.which = W_INIT;
+        this.it = 0;
+        this.ic = 0;
+    }
+
+    /** @inheritdoc */
+    moveNext() {
+
+        // Increment the index of tokens/comments.
+        if (this.which === W_INIT) {
+            this.it = this.firstIndex;
+            this.ic = search(this.comments, this.start);
+        } else if (this.which === W_TOKENS) {
+            this.it += 1;
+        } else {
+            this.ic += 1;
+        }
 
-        return get(first - (beforeCount || 0), first);
-    };
+        // Choose the current token from the token or the comment.
+        const t = (this.it < this.tokens.length) ? this.tokens[this.it] : null;
+        const c = (this.ic < this.comments.length) ? this.comments[this.ic] : null;
+
+        if (t && (!c || t.range[0] < c.range[0])) {
+            this.which = W_TOKENS;
+            this.current = t;
+        } else if (c) {
+            this.which = W_COMMENTS;
+            this.current = c;
+        } else {
+            this.current = null;
+        }
+
+        // Check the current token is before end.
+        return Boolean(this.current) && (this.end === -1 || this.current.range[1] <= this.end);
+    }
+
+    /** @inheritdoc */
+    movePrev() {
+
+        // Decrement the index of tokens/comments.
+        if (this.which === W_INIT) {
+            this.it = this.lastIndex;
+            this.ic = search(this.comments, this.end) - 1;
+        } else if (this.which === W_TOKENS) {
+            this.it -= 1;
+        } else {
+            this.ic -= 1;
+        }
+
+        // Choose the current token from the token or the comment.
+        const t = (this.it >= 0) ? this.tokens[this.it] : null;
+        const c = (this.ic >= 0) ? this.comments[this.ic] : null;
```

Comment:
The variable names `t` and `c` are a bit confusing. Maybe they should be something like `previousToken` and `previousComment` instead.